etc/schema/mysql/mysql.schema.sql

```diff
@@ -326,11 +326,12 @@ CREATE TABLE environment (
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=DYNAMIC;
 
 CREATE TABLE icingadb_instance (
-  id binary(16) NOT NULL COMMENT 'UUIDv4',
+  id binary(20) NOT NULL COMMENT 'sha1(environment.name + endpoint.name)',
```

Comment:
Would you use a composite primary key instead? Would be the only table doing this I guess.

Reply:
What for?

Reply:
I'm not sure what you want to say? Is it about the `id` column containing redundant information given the columns `environment_id` and `endpoint_name`? What would you do instead? The table needs some primary key over this information for how the rest of the code works.

Reply:
What _I'd do_? I'd leave the schema as-is.

The real questions: what _have you done_ and _what for_? You've:

1. Added a `binary(16)` random-value column
2. Removed a `binary(16)` random-value column
3. Added a `binary(20)` column with literally the same info as `environment_id`

Reply:
The real change is that the primary key is now deterministic. `startup_id` is just added as a safety feature to detect broken setups. Indeed, `id` currently hashes over the same information as `endpoint_id`, so this one could also be used as the primary key.

Reply:
I'd use id as startup_id and endpoint_id as id.

Reply:
The idea is that each instance deterministically generates a hopefully unique instance ID. That's what currently gets written into the `id` column. This ID identifies the row in `icingadb_instance` for which this instance is authoritative.

`startup_id` is then explicitly random in order to catch configurations where the "hopefully unique" assumption does not hold. It is generated randomly on each startup and then written to the database. On each heartbeat, Icinga DB will check if this ID is updated after it has started and if so, bail out.

Reply:
So there are... two instance ids?

Reply:
In a way, yes.

Reply:
I mean: two random instance ids?

Reply:
No, `id` has a deterministic value. This is used to prevent this table form growing indefinitely.

Reply:
I still don’t get why that whole concept should not fit into endpoint_id and id w/o changing them.

Reply:
It makes perfect sense if you know the history how I came up with this. :D

But I think you're right, keeping `icingadb_instance.id` as a random value and then using `environment_id` and `endpoint_id` to prune old entries should work as well. And this has the additional benefit that there isn't even a conflicting ID scenario that we have to prevent/detect.